a subchapter to Practical Robot

Vegebutcher opened the pull request:
this subchapter shows how to refine the scripts from the previous two exercises

Vegebutcher committed:
a subchapter to Practical Robot

WiktoriaPrusik commented on g1ant-language/refining-your-code.md:
```diff
@@ -0,0 +1,94 @@
+## Refining Your Code
+
+The code is never good enough. Of course, the golden rule says that if it works, don’t fix it — and we couldn’t agree more. But when you are practicing with programming, refinements are the part of the learning process. You learn new things and you can look back at your previous works to see if it’s possible to use your latest knowledge to optimize the code, make it clearer and more efficient.
```
golden rule not to refactor your code? I disagree

Vegebutcher replied:
GENERAL golden rule is: "if it ain't broke, don't fix it"

WiktoriaPrusik replied:
I don't think developers would agree with you, refactor is an important part of developing and without it "if it ain't broke, it most probably will be broke" it's because if your code works but no one knows why, it creates a situation in which only you can work on it and after a while you will find it extremely hard to do. Also, even a small change can do much harm... So it might be a GENERAL rule but not in programming :D

Vegebutcher replied:
Alright, I will make this sentence more precise in its generality :D
"Of course, the golden rule of life says that if it works, don’t fix it — and we couldn’t agree more."

WiktoriaPrusik replied:
hmm that wasn't what I was talking about hahahha I meant that it can remain the same if only you remove "and we couldn't agree more" oh come on! who agrees with that sentence? 😛

Vegebutcher replied:
ME! :smile:

Vegebutcher replied:
```suggestion
The code is never good enough. Of course, the golden rule of life says that if it ain't broke, don’t fix it — and we couldn’t agree more. But when you are practicing with programming, refinements are the part of the learning process. You learn new things and you can look back at your previous works to see if it’s possible to use your latest knowledge to optimize the code, make it clearer and more efficient.
```

WiktoriaPrusik commented on g1ant-language/refining-your-code.md:
````diff
@@ -0,0 +1,94 @@
+
+In the past two exercises, you used some basic commands and logic. In the first code, you can notice that there are many lines that could be reduced to just one, using a list variable and a loop:
+
+```G1
````
G1ANT*

Vegebutcher replied:
It's Typora's fault

Vegebutcher replied:
```suggestion
```G1ANT
```

WiktoriaPrusik commented on g1ant-language/refining-your-code.md:
```diff
@@ -0,0 +1,94 @@
+
+There is the `♥rownumber` variable which takes its value from the `for` loop, so it stays the same in all 6 rows of the procedure. How can you make the other two arguments change incrementally in the same line as the first one? By employing a list variable and another loop.
+
+Let’s start with the `♥cell` variable. Instead of creating a new variable for each cell, you could just create a list variable and add each cell’s content to this list as a value of a subsequent element. The `♥cell` list variable would look like this: `⟦cell1❚cell2❚cell3❚cell4❚cell5❚cell6⟧`. In order to move through the list elements, you would use the good old `foreach` loop or ordinary `for` lo ...
```
why did you create a list  ♥cell = ⟦cell1❚cell2❚cell3❚cell4❚cell5❚cell6⟧? what if you had 100 elements in that list? you should create an empty list instead:
`list.create type integer result ♥cells`
also I don't get why you inserted it inside `⟦⟧`. These characters are used to force parsing some value to some specified type for example: 
♥values = ⟦list:integer⟧0❚3❚10

WiktoriaPrusik commented on g1ant-language/refining-your-code.md:
````diff
@@ -0,0 +1,94 @@
+
+Before calling `GetValues` and `SetValues` procedures, you have to declare the `♥cell` variable and put it somewhere at the beginning of your script, after both `♥datafile` variables, for example:
+
+```G1
````
you have G1 everywhere, please correct it to G1ANT :)

Vegebutcher replied:
```suggestion
```G1ANT
```

WiktoriaPrusik replied:
you didn't change it everywhere ;)

Commit:
Update g1ant-language/refining-your-code.md

G1ANT

Commit:
Update g1ant-language/refining-your-code.md

G1ANT

Commit:
Update refining-your-code.md

included Wiktoria's suggestions regarding the list variable: list.create and list.add commands added

WiktoriaPrusik commented on g1ant-language/refining-your-code.md:
```diff
@@ -15,21 +15,21 @@ excel.getvalue row ♥rownumber colname f result ♥cell6
 There is the `♥rownumber` variable which takes its value from the `for` loop, so it stays the same in all 6 rows of the procedure. How can you make the other two arguments change incrementally in the same line as the first one? By employing a list variable and another loop.
 
-Let’s start with the `♥cell` variable. Instead of creating a new variable for each cell, you could just create a list variable and add each cell’s content to this list as a value of a subsequent element. The `♥cell` list variable would look like this: `⟦cell1❚cell2❚cell3❚cell4❚cell5❚cell6⟧`. In order to move through the list elements, you would use the good old `foreach` loop or ordinary `for` lo ...
+Let’s start with the `♥cells` variable. Instead of creating a new variable for each cell, you could just create a list variable and add each cell’s content to this list as a value of a subsequent element. The resulting `♥cells` list variable in this case would look like this: `cell1❚cell2❚cell3❚cell4❚cell5❚cell6`. In order to move through the list elements, you would use the good old `foreach` lo ...
```
you have changed your code but this list remained, so please update this description :)

Vegebutcher replied:
it's here for a reason: it shows how this list would look like. Again, it's an EXAMPLE, with EXAMPLE data representing cell values, not `cellX` variables or values.

WiktoriaPrusik replied:
so maybe add something like "but you can create an empty list instead" and then something like "you can do this using list.create command"

Vegebutcher replied:
"The **resulting** `♥cells` list variable in this case **would** look like this: "

key words: **resulting** and **would**

WiktoriaPrusik replied:
you don't explain how to use list.create command

WiktoriaPrusik commented on g1ant-language/refining-your-code.md:
```diff
@@ -15,21 +15,21 @@ excel.getvalue row ♥rownumber colname f result ♥cell6
 ♥datafile1 = ‴♥environment⟦USERPROFILE⟧\Documents\Data\data.xlsx‴
 ♥datafile2 = ‴♥environment⟦USERPROFILE⟧\Documents\Data\data2.xlsx‴
-♥cell = ⟦cell1❚cell2❚cell3❚cell4❚cell5❚cell6⟧
+list.create ♥cells
```
this is incorrect, have you tried running such code? :p 
should be: `list.create result ♥cells`

Vegebutcher replied:
```suggestion
list.create result ♥cells
```

WiktoriaPrusik commented on g1ant-language/refining-your-code.md:
````diff
@@ -15,21 +15,21 @@ excel.getvalue row ♥rownumber colname f result ♥cell6
+list.create ♥cells
 ```
 
 The `for` loop, inside which the procedures are called, gives sequential row numbers to be used by the `row` argument and its  `♥rownumber` variable. In each row you want to process six cells in six columns. Therefore, for both procedures you have to set up another counter in the `for` loop, which will count from 1 to 6 and pass this value both to a column number in the `colindex` argument and a  ...
````
if you rename your variable to cells, then it this variable should be renamed everywhere :)

Vegebutcher replied:
I did rename it, probably didn't notice this one

Vegebutcher replied:
```suggestion
The `for` loop, inside which the procedures are called, gives sequential row numbers to be used by the `row` argument and its  `♥rownumber` variable. In each row you want to process six cells in six columns. Therefore, for both procedures you have to set up another counter in the `for` loop, which will count from 1 to 6 and pass this value both to a column number in the `colindex` argument and a cell index in the `♥cells` list variable. Let’s name this counter `♥col` and insert the loop into each procedure:
```

Commit:
Update g1ant-language/refining-your-code.md

cell -> cells

Commit:
Update g1ant-language/refining-your-code.md

result arg added

Commit:
Update g1ant-language/refining-your-code.md

ain't broke

Vegebutcher commented:
@ChrisPrusik, last chance!

Commit:
Update refining-your-code.md

Headers changed (elevated), one relative link corrected

Commit:
Rename refining-your-code.md to 03-refining-your-code.md

filename changed

Commit:
Rename g1ant-language/03-refining-your-code.md to g1ant-language/practical-robot/03-refining-your-code.md

folder changed
